source/extensions/io_socket/user_space/file_event_impl.cc

```diff
@@ -0,0 +1,76 @@
+#include "extensions/io_socket/user_space/file_event_impl.h"
+
+#include <cstdint>
+
+#include "common/common/assert.h"
+
+#include "extensions/io_socket/user_space/io_handle.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace IoSocket {
+namespace UserSpace {
+
+FileEventImpl::FileEventImpl(Event::Dispatcher& dispatcher, Event::FileReadyCb cb, uint32_t events,
+                             IoHandle& io_source)
+    : schedulable_(dispatcher.createSchedulableCallback([this, cb]() {
+        auto ephemeral_events = event_listener_.getAndClearEphemeralEvents();
+        ENVOY_LOG(trace, "User space event {} invokes callbacks on events = {}",
+                  static_cast<void*>(this), ephemeral_events);
+        cb(ephemeral_events);
+      })),
+      io_source_(io_source) {
+  setEnabled(events);
+}
+
+void FileEventImpl::activate(uint32_t events) {
+  // Only supported event types are set.
+  ASSERT((events & (Event::FileReadyType::Read | Event::FileReadyType::Write |
+                    Event::FileReadyType::Closed)) == events);
+  event_listener_.onEventActivated(events);
+  schedulable_->scheduleCallbackNextIteration();
+}
+
+void FileEventImpl::setEnabled(uint32_t events) {
+  // Only supported event types are set.
+  ASSERT((events & (Event::FileReadyType::Read | Event::FileReadyType::Write |
+                    Event::FileReadyType::Closed)) == events);
+  event_listener_.clearEphemeralEvents();
+  event_listener_.setEnabledEvents(events);
+  bool was_enabled = schedulable_->enabled();
+  // Recalculate activated events.
+  uint32_t events_to_notify = 0;
+  if ((events & Event::FileReadyType::Read) && io_source_.isReadable()) {
```

Comment:
if ((events & Event::FileReadyType::Read) && (io_source_.isReadable() || io_source_.isPeerShutDownWrite()) {

Reply:
Aha, great catch! Fixing...

Reply:
Actually, in the interface I add the comment that peerShutdownWrite() imply `isReadable()`. But it's nice to have explicit calls.

```
  /**
   * @return true if the pending receive buffer is not empty or read_end is set.
   */
  virtual bool isReadable() const PURE;
```